generators/E906LegacyVtxGen/SQPrimaryVertexGen.C

```diff
@@ -0,0 +1,270 @@
+      geoManager->SetCurrentDirection(0., 0., 1.);
+
+      TGeoNode* node = geoManager->FindNextBoundary(600.);
+      double z_step  = geoManager->GetStep();
+
+      if(z_step > eps)
+	{
+	  z_curr += z_step;
+	  z[++nBoundaries] = z_curr; 
+	}
+      else
+	{
+	  z_curr += eps;
+	}
+    }
+
+  //use the z boundaries to create interactibles
+  for(int i = 0; i < nBoundaries; ++i)
+    {
+      double z_center = 0.5*(z[i] + z[i+1]);
+      if(z_center > 503.) continue;
+ 
+      SQBeamlineObject obj(geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetMaterial());
+      TString name = geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetName();
+      obj.z_up = z[i];
+      obj.z_down = z[i+1];
+      obj.z0 = z_center;
+      obj.length = z[i+1] - z[i];
+    
+      //TString name = obj.name;    
+      //std::cout<<" Name of beam volume "<<name<<std::endl;  
+
+      if (_targetOnlyMode){
+	if(! name.Contains("Target")) continue;
+      }
+
+      if (_dumpOnlyMode){
+	if(! name.Contains("fmag")) continue;
+      }
+
+      prof->interactables.push_back(obj);
+    }
+
+  //calculate all related properties
+  prof->calcProb();
+
+  // for(int i = 0; i < prof->nPieces; ++i)
+  // {
+  //   std::cout << prof->interactables[i] << std::endl;
+  // }
+  // return 0;
+}
+
+
+TVector3 SQPrimaryVertexGen::generateVertex()
+{
+  if(!inited) init();
+
+  double xvtx, yvtx, zvtx;
+  generateVtxPerp(xvtx, yvtx);
+
+  bool inTarget = ((xvtx-targetX0)*(xvtx-targetX0)/targetSX/targetSX + (yvtx-targetY0)*(yvtx-targetX0)/targetSY/targetSY) < 1.;
+  MaterialProfile* activeProf;
+  if(!inTarget)
+    {
+      
+      activeProf = new MaterialProfile;
+      fillMaterialProfile(activeProf, xvtx, yvtx);
+    }
+  else
+    {
+      activeProf = defaultMatProf;
+    }
+
+  int index = activeProf->findInteractingPiece(rndm.Rndm());
+  zvtx = activeProf->interactables[index].getZ(rndm.Rndm());
+  pARatio = activeProf->interactables[index].protonPerc();
+  probSum = activeProf->probSum;
+
+  if(!inTarget) delete activeProf;
+
+  return TVector3(xvtx, yvtx, zvtx);
+}
+
+void SQPrimaryVertexGen::generateVtxPerp(double& x, double& y)
+{
+  beamProfile->GetRandom2(x, y);
+}
+
+
+//!Function for beam profile (double gaussian at core and 1/r at tail)
+double SQPrimaryVertexGen::funcBeamProfile(double* val, double* par)
+{
+  double x   = val[0];
+  double y   = val[1];
+  double x_c = par[0];
+  double x_s = par[1];
+  double y_c = par[2];
+  double y_s = par[3];
+  
+  const double r_pipe = 5.0; // in cm
+  const double r_boun = 3.0; // normalized boundary, where gaus changes to 1/r
+  if (pow(x,2) + pow(y,2) > pow(r_pipe,2)) { // not "x-x_c" nor "y-y_c" here
+    return 0.0;
+  }
+
+  //@ for target only mode
+  recoConsts* rc = recoConsts::instance();   
+  bool inTarget = (pow(x-rc->get_DoubleFlag("X0_TARGET"),2)/pow(rc->get_DoubleFlag("RX_TARGET"),2 )+ pow(y-rc->get_DoubleFlag("Y0_TARGET"),2)/pow(rc->get_DoubleFlag("RY_TARGET"),2)) < 1.; 
```

Comment:
This part seems unnatural.  In physics, the beam profile should not depend on any condition of target/dump materials.  In technique, this part makes the `_targetOnlyMode` variable static.

Why do you need this part?  I expect the target-only mode can be fulfilled by the target side, i.e. `fillMaterialProfile()` only.

Reply:
Yes. But if we choose target only condition in fillMaterialProfile(), it will cause crash because of not finding any material for vertex (x,y) outside the target for the event. Also, this condition in beam profile will only generate (X,Y) from the part where the target can interact without having to modify the actual gaussian function (inner core and outer tail). 

But for FMAG, this won't happen since the beam profile we defined falls inside the range of the width of FMAG.

Reply:
That sounds a problem specific to `fillMaterialProfile()`.  It should check and properly handle the case of no material for any geometry configuration.  Can you add such a check?

Reply:
I will try to look into it. Thank you Kenichi.

Reply:
I will give it a try to not inheriting from the SubsystemReco

Reply:
Yes, now it is not inheriting from the SubsysReco. Thank you.